File: src/api/share.rs
//! `/api/v1/share/*` — public share-link surface per Phase 1.g.
//!
//! Five routes split across the same auth-vs-public boundary the
//! streaming module uses:
//!
//! - **Mint** (`POST /api/v1/profiles/{profile_id}/playlists/{playlist_id}/share`):
//!   JWT-authed. Verifies tenant ownership of the playlist, then
//!   atomically generates (or returns the existing) opaque token via
//!   `db::share::mint_or_get_token`. Idempotent — a second call for
//!   the same playlist returns the same token rather than rotating
//!   it.
//! - **Revoke** (`DELETE /api/v1/profiles/{profile_id}/playlists/{playlist_id}/share`):
//!   JWT-authed. Sets `share_token = NULL`, instantly closing any
//!   public URL pointing at this playlist.
//! - **Mint by canonical**
//!   (`POST /api/v1/share/playlists/by-canonical/{profile_canonical_id}/{playlist_canonical_id}`):
//!   JWT-authed. Same semantics as `mint`, but keyed on the desktop's
//!   canonical UUIDs (Phase 1.g.0). The desktop never sees the
//!   server-side BIGSERIAL ids the apply pipeline assigns, so this
//!   variant skips the lookup round-trip the desktop would otherwise
//!   need to translate canonical → server id before calling the
//!   classic endpoint.
//! - **Revoke by canonical**
//!   (`DELETE /api/v1/share/playlists/by-canonical/{profile_canonical_id}/{playlist_canonical_id}`):
//!   Mirror of revoke for the canonical-id surface.
//! - **Public read** (`GET /api/v1/share/playlists/{token}`): NOT
//!   behind the JWT middleware — the token IS the auth. A miss (no
//!   row matches the token) returns 404 with no body so an attacker
//!   can't distinguish "revoked" from "never minted".
//!
//! Wire format intentionally minimal: name + description + cover +
//! brand tokens + timestamps. Track list is NOT returned today
//! because the server doesn't materialise `playlist_track` yet
//! (the desktop is still the source of truth for the join). When
//! Phase 1.g.2 brings server-side materialisation, this DTO grows a
//! `tracks: Vec<PublicTrack>` field without a wire-break.

use axum::{
    extract::{Extension, Path, State},
    http::StatusCode,
    response::IntoResponse,
    Json,
};
use serde::Serialize;
use utoipa::ToSchema;
use utoipa_axum::{router::OpenApiRouter, routes};

use crate::{middleware::UserId, AppState};

/// Body of the mint response. `token` is the opaque ID the web
/// client uses for the public route (`/p/<token>` on
/// waveflow-web). The server does NOT return the full URL because
/// it doesn't know the web origin — clients combine this token
/// with their persisted `app_setting['app.waveflow_web_url']` to
/// build the shareable link.
#[derive(Debug, Serialize, ToSchema)]
pub struct MintResponse {
    /// Opaque URL-safe token (32 alphanumeric chars). Stable for
    /// the lifetime of the link — a second mint call returns the
    /// same value rather than rotating.
    pub token: String,
}

/// Public preview of a shared playlist. Mirrors the desktop's
/// `Playlist` DTO minus the `profile_id` (the share owner isn't
/// exposed to anonymous viewers) and minus the smart-playlist
/// machinery (smart playlists can't be shared — they materialise
/// per-device).
#[derive(Debug, Serialize, ToSchema)]
pub struct PublicPlaylistResponse {
    pub id: i64,
    pub name: String,
    pub description: Option<String>,
    pub color_id: String,
    pub icon_id: String,
    /// BLAKE3 hash of the cover image in the shared metadata cache.
    /// `None` until the artwork pipeline ships on the server.
    pub cover_hash: Option<String>,
    pub created_at: i64,
    pub updated_at: i64,
    /// Track list. Always empty today — server-side
    /// `playlist_track` materialisation is Phase 1.g.2. The field
    /// is present so a future server release can populate it
    /// without breaking the wire shape.
    pub tracks: Vec<PublicTrack>,
}

/// Placeholder shape for the track list — kept minimal so the
/// initial wire contract doesn't pre-commit to a richer DTO that
/// would need re-negotiating once the join lands server-side.
#[derive(Debug, Serialize, ToSchema)]
pub struct PublicTrack {
    pub title: String,
    pub artist: Option<String>,
    pub duration_ms: i64,
}

pub fn auth_router(state: AppState) -> OpenApiRouter {
    OpenApiRouter::new()
        .routes(routes!(mint_share_token))
        .routes(routes!(revoke_share_token))
        .routes(routes!(mint_share_token_by_canonical))
        .routes(routes!(revoke_share_token_by_canonical))
        .with_state(state)
}

pub fn public_router(state: AppState) -> OpenApiRouter {
    OpenApiRouter::new()
        .routes(routes!(get_public_playlist))
        .with_state(state)
}

#[utoipa::path(
    post,
    path = "/api/v1/profiles/{profile_id}/playlists/{playlist_id}/share",
    tag = "share",
    params(
        ("authorization" = String, Header, description = "Bearer JWT issued by Better Auth"),
        ("profile_id" = i64, Path, description = "Owning profile id"),
        ("playlist_id" = i64, Path, description = "Playlist to publish"),
    ),
    responses(
        (status = 200, description = "Token minted or echoed back if one already existed", body = MintResponse),
        (status = 401, description = "Missing or invalid bearer token"),
        (status = 404, description = "Playlist not found in the requested profile or not owned by the caller"),
        (status = 500, description = "Database or internal failure"),
    ),
)]
async fn mint_share_token(
    State(state): State<AppState>,
    Extension(UserId(user_id)): Extension<UserId>,
    Path((profile_id, playlist_id)): Path<(i64, i64)>,
) -> impl IntoResponse {
    match crate::db::share::mint_or_get_token(&state.db, user_id, profile_id, playlist_id).await {
        Ok(Some(token)) => (StatusCode::OK, Json(MintResponse { token })).into_response(),
        Ok(None) => (StatusCode::NOT_FOUND, "playlist not found").into_response(),
        Err(err) => {
            tracing::error!(error = %err, user_id, profile_id, playlist_id, "share mint failed");
            (StatusCode::INTERNAL_SERVER_ERROR, "mint failed").into_response()
        }
    }
}

#[utoipa::path(
    delete,
    path = "/api/v1/profiles/{profile_id}/playlists/{playlist_id}/share",
    tag = "share",
    params(
        ("authorization" = String, Header, description = "Bearer JWT issued by Better Auth"),
        ("profile_id" = i64, Path, description = "Owning profile id"),
        ("playlist_id" = i64, Path, description = "Playlist whose link should be closed"),
    ),
    responses(
        (status = 204, description = "Token cleared (idempotent — also returned if the link was already private)"),
        (status = 401, description = "Missing or invalid bearer token"),
        (status = 404, description = "Playlist not found in the requested profile or not owned by the caller"),
        (status = 500, description = "Database or internal failure"),
    ),
)]
async fn revoke_share_token(
    State(state): State<AppState>,
    Extension(UserId(user_id)): Extension<UserId>,
    Path((profile_id, playlist_id)): Path<(i64, i64)>,
) -> impl IntoResponse {
    match crate::db::share::revoke_token(&state.db, user_id, profile_id, playlist_id).await {
        Ok(true) => StatusCode::NO_CONTENT.into_response(),
        Ok(false) => (StatusCode::NOT_FOUND, "playlist not found").into_response(),
        Err(err) => {
            tracing::error!(error = %err, user_id, profile_id, playlist_id, "share revoke failed");
            (StatusCode::INTERNAL_SERVER_ERROR, "revoke failed").into_response()
        }
    }
}

#[utoipa::path(
    post,
    path = "/api/v1/share/playlists/by-canonical/{profile_canonical_id}/{playlist_canonical_id}",
    tag = "share",
    params(
        ("authorization" = String, Header, description = "Bearer JWT issued by Better Auth"),
        ("profile_canonical_id" = String, Path, description = "Desktop profile's canonical UUID"),
        ("playlist_canonical_id" = String, Path, description = "Desktop playlist's canonical UUID"),
    ),
    responses(
        (status = 200, description = "Token minted or echoed back", body = MintResponse),
        (status = 401, description = "Missing or invalid bearer token"),
        (status = 404, description = "Profile or playlist not found (or not owned by the caller)"),
        (status = 500, description = "Database or internal failure"),
    ),
)]
async fn mint_share_token_by_canonical(
    State(state): State<AppState>,
    Extension(UserId(user_id)): Extension<UserId>,
    Path((profile_canonical_id, playlist_canonical_id)): Path<(String, String)>,
) -> impl IntoResponse {
    match crate::db::share::mint_or_get_token_by_canonical(
        &state.db,
        user_id,
        &profile_canonical_id,
        &playlist_canonical_id,
    )
    .await
    {
        Ok(Some(token)) => (StatusCode::OK, Json(MintResponse { token })).into_response(),
        Ok(None) => (StatusCode::NOT_FOUND, "playlist not found").into_response(),
        Err(err) => {
            tracing::error!(
                error = %err,
                user_id,
                profile_canonical_id = %profile_canonical_id,
                playlist_canonical_id = %playlist_canonical_id,
                "share mint (by canonical) failed",
            );
            (StatusCode::INTERNAL_SERVER_ERROR, "mint failed").into_response()
        }
    }
}

#[utoipa::path(
    delete,
    path = "/api/v1/share/playlists/by-canonical/{profile_canonical_id}/{playlist_canonical_id}",
    tag = "share",
    params(
        ("authorization" = String, Header, description = "Bearer JWT issued by Better Auth"),
        ("profile_canonical_id" = String, Path, description = "Desktop profile's canonical UUID"),
        ("playlist_canonical_id" = String, Path, description = "Desktop playlist's canonical UUID"),
    ),
    responses(
        (status = 204, description = "Token cleared (idempotent)"),
        (status = 401, description = "Missing or invalid bearer token"),
        (status = 404, description = "Profile or playlist not found (or not owned by the caller)"),
        (status = 500, description = "Database or internal failure"),
    ),
)]
async fn revoke_share_token_by_canonical(
    State(state): State<AppState>,
    Extension(UserId(user_id)): Extension<UserId>,
    Path((profile_canonical_id, playlist_canonical_id)): Path<(String, String)>,
) -> impl IntoResponse {
    match crate::db::share::revoke_token_by_canonical(
        &state.db,
        user_id,
        &profile_canonical_id,
        &playlist_canonical_id,
    )
    .await
    {
        Ok(true) => StatusCode::NO_CONTENT.into_response(),
        Ok(false) => (StatusCode::NOT_FOUND, "playlist not found").into_response(),
        Err(err) => {
            tracing::error!(
                error = %err,
                user_id,
                profile_canonical_id = %profile_canonical_id,
                playlist_canonical_id = %playlist_canonical_id,
                "share revoke (by canonical) failed",
            );
            (StatusCode::INTERNAL_SERVER_ERROR, "revoke failed").into_response()
        }
    }
}

#[utoipa::path(
    get,
    path = "/api/v1/share/playlists/{token}",
    tag = "share",
    params(
        ("token" = String, Path, description = "Opaque share token minted by POST /share"),
    ),
    responses(
        (status = 200, description = "Public preview of the shared playlist", body = PublicPlaylistResponse),
        (status = 404, description = "Token unknown (never minted or revoked)"),
        (status = 500, description = "Database or internal failure"),
    ),
)]
async fn get_public_playlist(
    State(state): State<AppState>,
    Path(token): Path<String>,
) -> impl IntoResponse {
    match crate::db::share::fetch_public_by_token(&state.db, &token).await {
        Ok(Some((
            id,
            name,
            description,
            color_id,
            icon_id,
            cover_hash,
            created_at,
            updated_at,
        ))) => (
            StatusCode::OK,
            Json(PublicPlaylistResponse {
                id,
                name,
                description,
                color_id,
                icon_id,
                cover_hash,
                created_at,
                updated_at,
                tracks: Vec::new(),
            }),
        )
            .into_response(),
        Ok(None) => StatusCode::NOT_FOUND.into_response(),
        Err(err) => {
            tracing::error!(error = %err, "share public lookup failed");
            (StatusCode::INTERNAL_SERVER_ERROR, "lookup failed").into_response()
        }
    }
}

File: web/src/server-fns/share.ts
// Server functions for the public `/api/v1/share/*` surface (Phase
// 1.g.2 — the read side of playlist sharing). Mint + revoke stay on
// the desktop / authed callers; the web only consumes the read
// endpoint that powers the `/p/$token` route.

import { notFound } from '@tanstack/react-router'
import { createServerFn } from '@tanstack/react-start'
import { waveflowFetchPublic, WaveflowServerError } from '@/lib/server/waveflow-server'

/**
 * Minimal placeholder track shape mirroring the server's
 * `PublicTrack` DTO. Empty for every playlist today (server-side
 * `playlist_track` materialisation is a separate Phase 1.g.x); the
 * field exists so a future server release can populate it without
 * a wire-break.
 */
export interface PublicTrack {
  title: string
  artist: string | null
  duration_ms: number
}

/**
 * Public preview of a shared playlist. Mirrors the server's
 * `PublicPlaylistResponse` — name + description + cover + brand
 * tokens + timestamps + the tracks list (populated by Phase 1.j
 * desktop emitters; older clients leave it empty and the route
 * renders the "no preview yet" fallback). Notably absent:
 * `profile_id` and any other tenant-identifying fields. Same DTO
 * drives the `<meta>` tag rendering on `/p/$token`.
 */
export interface PublicPlaylist {
  id: number
  name: string
  description: string | null
  color_id: string
  icon_id: string
  cover_hash: string | null
  created_at: number
  updated_at: number
  tracks: PublicTrack[]
}

/**
 * The server-fn returns a discriminated union for the "reached
 * waveflow-server" path. "Playlist not found" is signalled out-
 * of-band via `throw notFound()` — the router catches it,
 * renders the route's `notFoundComponent` and emits a real HTTP
 * 404. Going through the union for the 404 case would force the
 * TanStack RPC client to swallow a non-OK HTTP response, which
 * it treats as a thrown error (see `serverFnFetcher.ts:188`).
 */
export type PublicPlaylistResult =
  | { kind: 'ok'; playlist: PublicPlaylist }
  | { kind: 'error'; message: string }

/**
 * Reject obviously malformed tokens at the boundary so a hand-
 * crafted server-fn payload can't trick us into proxying a path
 * like `/api/v1/share/playlists/../whatever` to waveflow-server.
 *
 * The server mints exactly 32 alphanumeric characters
 * (`rand::distributions::Alphanumeric`); we mirror that exactly
 * on the client. Tight enough to block any URL-meaningful
 * character (slash, dot, query, hash, percent) and any deviation
 * from the server's mint format.
 *
 * The check is consulted by the handler — not the input validator
 * — so a malformed token resolves to `{ kind: 'not_found' }` and
 * the route renders the friendly empty state instead of crashing
 * into the framework's error boundary.
 */
export function isWellShapedToken(value: unknown): value is string {
  return typeof value === 'string' && /^[A-Za-z0-9]{32}$/.test(value)
}

export const getPublicPlaylist = createServerFn({ method: 'GET' })
  .inputValidator((value: unknown): string => {
    // Pass through as a string — shape validation happens in the
    // handler so a bad token can surface as a 'not_found' result
    // rather than a thrown error the route can't render.
    return typeof value === 'string' ? value : ''
  })
  .handler(async ({ data: token }): Promise<PublicPlaylistResult> => {
    // Throw `notFound()` for the 404 cases — TanStack serialises
    // it through the RPC layer as a recognised payload (vs.
    // throwing every non-OK HTTP as an Error), so the client
    // loader sees a real `NotFoundError` and the router routes
    // to `notFoundComponent` with a proper SSR 404 status.
    if (!isWellShapedToken(token)) {
      throw notFound()
    }
    try {
      const playlist = await waveflowFetchPublic<PublicPlaylist>(
        `/api/v1/share/playlists/${encodeURIComponent(token)}`,
      )
      return { kind: 'ok', playlist }
    } catch (err) {
      if (err instanceof WaveflowServerError && err.status === 404) {
        throw notFound()
      }
      console.error('[server-fn] getPublicPlaylist failed:', err)
      return {
        kind: 'error',
        message: 'Could not reach waveflow-server. Please try again.',
      }
    }
  })
